### ISSUE TITLE
Execute ad hoc commands using ansible-container?

### ISSUE THREAD

**Issue by containscafeine** [ansible/ansible-container#204]
##### ISSUE TYPE

 - Feature Idea

It would be pretty sweet if we could execute ad hoc commands on all/some of the running containers.

We can leverage docker compose's labels, Kubernetes' labels and selectors for this, to execute on a subset of running services.

I am suggesting something very similar to ansible's ad hoc command line, like `ansible-container exec --tag db -m shell -a "tail /var/log/db/error.log" ` or `ansible-container exec --host webserver -m shell -a "tail /var/log/httpd/access.log"`

I have not thought this through, but any thoughts? Does this make sense?

**Comment by chouseknecht** [comment]
I have not fully formed an opinion on this, so I'm thinking out loud and asking a few questions. I would like to see it spec'd out a bit more to ensure there is agreement on the purpose and scope of the feature (particularly the options), so hopefully my questions will help to that end.

My first thought is to consider the alternative to adding this feature, which I believe is to execute the following commands:

```
$ ansible-container run -d
$ docker ps
... List of container names ...
$ docker exec <container name> <command>
```

That's not horrible. And, if you know your list of service names and how container names are constructed, you can skip running `docker ps`. It also means the `docker exec` part can be generated by a test script.

Questions: 

* Given the alternative, as stated above, do the benefits outweigh the cost of writing and maintaining the code for this feature? 

* Is there a clear use case for the feature? The use case behind `ansible-container run` is providing a way to test the built images. In order to test images, we need to create containers and then make assertions, and of course `run` makes that possible. Perhaps `ansible-container exec` is an additional testing tool?

* Is there a use case where running this command against multiple services makes sense? I'm making the assumption that the purpose of --tag is to run the command against multiple services, which means the code will have to deal with the output and results from multiple services, which seems like it could get complicated. 

* What happens if scaling is implemented? Don't know that it would be or should be implemented, but one day it might be. (I believe there is actually an open issue for it.) If it turned out that scaling did get implemented, and we had 5 instances of a single service running, how would `ansible-container exec` behave?

**Comment by containscafeine** [comment]
> My first thought is to consider the alternative to adding this feature, which I believe is to execute the following commands
> ```bash
$ ansible-container run -d
> $ docker ps
> ... List of container names ...
> $ docker exec <container name> <command>```

So the use case is that, if my application definition in container.yml has 10 services, out of which 5 are labelled as `db` and I want the database logs from each one of them, then instead of having to `docker exec` into every database container, I can just fire one command `ansible-container exec --tag db tail /var/log/...` and have the logs from every database container.

Also, since we are implementing `ansible-container run --provider k8s/oc` this will also translate to those providers as well. Both of them have labels implemented, and we can use selectors to execute commands to those too.

This is a very similar use case to using ansible ad hoc commands.

> Given the alternative, as stated above, do the benefits outweigh the cost of writing and maintaining the code for this feature?

Umm, I see myself using this a lot once this is implemented, since this lets me execute commands to a group of containers on the fly.

> Is there a clear use case for the feature? The use case behind ansible-container run is providing a way to test the built images. In order to test images, we need to create containers and then make assertions, and of course run makes that possible. Perhaps ansible-container exec is an additional testing tool?

Makes sense, but the way I look at ansible-container is an abstraction tool that lets me develop without having to worry about the provider in the background. Something like the vagrant workflow? It gives me enough to get my development environment up and running to deploy my app, but abstracts all the provider specific intricacies away from me, so I get a uniform workflow throughout without having to switch tools.
In this case, I want more than just test my images, I want to iterate on my code and rebuild, redeploy over and over, and executing ad hoc commands does help me iterate faster, helps with troubleshooting and what not. Makes sense?

> Is there a use case where running this command against multiple services makes sense? I'm making the assumption that the purpose of --tag is to run the command against multiple services, which means the code will have to deal with the output and results from multiple services, which seems like it could get complicated.

Well, same with ad hoc commands in ansible, it helps in making changes on the fly and troubleshooting my application, and the code should just run `docker exec` over the selected containers and execute the command, in a loop. The code can be dumb and just display the output, does not have to process it, maybe?

> What happens if scaling is implemented? Don't know that it would be or should be implemented, but one day it might be. (I believe there is actually an open issue for it.) If it turned out that scaling did get implemented, and we had 5 instances of a single service running, how would ansible-container exec behave?

In that case, the provider should help us with the tagged resources and we will ask the provider to execute the commands for us. `kubectl exec` etc should help!

---

In short, this would help me iterate faster, troubleshoot better, make changes instantaneously to a group of containers.

Thanks for taking a look at this! :)

**Comment by j00bar** [comment]
This can currently be done using `ansible -c docker`, no?

**Comment by containscafeine** [comment]
@j00bar I believe yes, but I was proposing using docker/provider's labels to query the containers instead of the tags in ansible playbooks.

**Comment by sara-02** [comment]
@j00bar, @chouseknecht this looks interesting. I will give it a shot during PyCon India, and send a PR.

**Comment by gregdek** [comment]
@sara-02 any luck?

**Comment by sara-02** [comment]
@gregdek None as of now.
